File: LowestCommonAncestorOfBT.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


"""
Approach: DFS

TC: O(n). We may traverse all the nodes
SC: O(n). We are storing in list

n = no of nodes
"""
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root == None:
            return root
        
        path1 = [] # path for p
        path2 = [] # path for q
        
        self.backtrack(root, p, path1)
        self.backtrack(root, q, path2)
        n1 = len(path1)
        n2 = len(path2)
        
        i = 0
        while i < min(n1,n2):
            if path1[i] != path2[i]:
                break
            i += 1
        
        return path1[i - 1]
        
    def backtrack(self, root, node, path):
        if root == None:
            return
        
        path.append(root)
        
        # Instead of comparing values its better to compare references, because even we get duplicated values the code should work
        if root == node:
            return
        
        self.backtrack(root.left, node, path)
        
        """
        We are adding this because without it the recursion will keep going even after we got the path.
        It is performing correctly before the value being found but after the path is found we need stop           it there. So after the path is found, keep returning until recurssion stack is empty.  
        
        Basically this condition makes the removing stop once path is found.    
        """
        if path[-1] == node:
            return
        
        self.backtrack(root.right, node, path)
        if path[-1] == node:
            return
        
        path.pop()
        
        

File: LowestCommonAncestorOfBT.java
// top down recursion
public class LowestCommonAncestorOfBT {
    List<TreeNode> pathp;
    List<TreeNode> pathq;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // null case
        if(root==null) return null;
        helper(root,p,q,new ArrayList<>());
        for(int i=0;i<pathp.size();i++){
            if(pathp.get(i)!=pathq.get(i)){
                return pathp.get(i-1);
            }
        }
        return null;
    }
    private void helper(TreeNode root, TreeNode p, TreeNode q, List<TreeNode> path){
        // base case
        if(root==null) return;

        path.add(root); // action
        if(root==p){
            pathp = new ArrayList<>(path);
            pathp.add(root);
        }
        if(root==q){
            pathq = new ArrayList<>(path);
            pathq.add(root);
        }

        // recurse
        helper(root.left,p,q,path);
        helper(root.right,p,q,path);

        // backtrack
        path.remove(path.size()-1);
    }
}

// TC - O(n)
// SC - O(h)


// bottom up recursion
public class LowestCommonAncestorOfBT {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root==null || root==p || root==q) return root;
        TreeNode left = lowestCommonAncestor(root.left,p,q);
        TreeNode right = lowestCommonAncestor(root.right,p,q);

        if(left==null && right==null) return null;
        else if(left!=null && right==null) return left;
        else if(left==null && right!=null) return right;
        else return root;
    }
}

// TC - O(n)
// SC - O(h)